sycl/source/detail/scheduler/scheduler.hpp

```diff
@@ -820,6 +849,22 @@ class Scheduler {
   // scheduler. If program is not correct and doesn't have necessary sync point
   // then warning will be issued.
   std::unordered_map<stream_impl *, StreamBuffers *> StreamBuffersPool;
+
+  /// Matches SYCL memory objects to attached resources.
+  /// TODO: On ABI break this could be made part of SYCLMemObjT instead.
+  std::unordered_map<const SYCLMemObjI *,
+                     std::vector<std::shared_ptr<const void>>>
+      m_MemObjLifetimeAttachedResources;
```

Comment:
While it's convenient, not sure it's conceptually a proper place for keeping such information even for USM allocations because it seems there is not related to what Scheduler does.  Probably we could introduce a new global state(e.g. MemoryManager stateful) for such memory allocations or submit some cleanup host task(but it will add more overhead).

Reply:
Scheduler seemed to be the best option of the current singletons, but I did stare at MemoryManager for a while hoping it was also a singleton. I don't mind adding a new singleton, though a benefit of having it in the scheduler is it allowing us to do early cleanups in the future, in case we think of places where we know it is safe to let go of additional resources (maybe a wait on an event for a reduction?). That said, a MemoryManager-like singleton would probably be reachable from the scheduler anyway though.

Reply:
Another place to consider is a `context`. We could reuse these buffers/USM allocations for reductions and clean up them during `queue::wait` or `event::wait`.

Reply:
Good idea! `free` has the context as well, so it would make perfect sense. Would also prevent the case where a user-side leak could cause unexpected problems. I will move it to `context` once tests are done.